vllm/distributed/communication_op.py

```diff
@@ -23,18 +36,19 @@ def tensor_model_parallel_all_reduce(input_: torch.Tensor) -> torch.Tensor:
     TLDR: always assume this function modifies its input, but use the return
     value as the output.
     """
-    from vllm.distributed.device_communicators import pynccl_utils
     from vllm.distributed.device_communicators.custom_all_reduce import (
         custom_all_reduce)
+    from vllm.distributed.parallel_state import _TP_PYNCCL_COMMUNICATOR
```

Comment:
> Why don't we import these at the top of the module?

Because this variable is `None` when the module gets initialized. Import these at the top of the module will get `None`. It will not be updated as the variable in the `vllm.distributed.parallel_state`.

> the variable is used outside the module where it's defined

I actually want to merge `parallel_state` and `communication_op` . `communication_op` needs to access many private variable in `parallel_state` . That's another story though.

Reply:
Why don't we use a get method then?

Reply:
a get method is doable. however, if the majority callers of these getter method come from `communication_op`, it's better to simply merge two files, imo.

Reply:
Maybe yes. But if we do not merge the files in this PR, I think using the getter method is a good idea. The current code has two problems: 1) It must be lazily imported for the reason you mentioned above, and 2) the variable name starts with `_`. I believe using the getter method is a simple solution to the two problems.

Reply:
I don't want this PR to be super complicated. I will merge the files in the next days, let's don't add new user-facing getter functions just for several days and remove them later 🤣

Reply:
Honestly, I think in this case we should add the getter function even if there's a change that it can be deleted just after several days. This is because we don't have a concrete timeline for the next PR you are thinking of. It can be delayed or de-prioritized any way. I believe adding the getter function doesn't hurt us.

Reply:
Added in [70a7e26](https://github.com/vllm-project/vllm/pull/4591/commits/70a7e2679b477e02c6559c4c47f9368d7c728ad3) .